tools/PI/DevHome.PI/Helpers/ExternalTool.cs

```diff
@@ -120,59 +124,121 @@ private async void GetIcons()
-                {
-                    argumentVariables.Add("pid", pid.Value);
-                }
-
-                if (hwnd.HasValue)
-                {
-                    argumentVariables.Add("hwnd", (int)hwnd.Value);
-                }
+                argumentVariables.Add("pid", pid.Value);
+            }
 
-                parsedArguments = ReplaceKnownVariables(Arguments, argumentVariables);
+            if (hwnd.HasValue)
+            {
+                argumentVariables.Add("hwnd", (int)hwnd.Value);
             }
 
+            parsedArguments = ReplaceKnownVariables(Arguments, argumentVariables);
+        }
+
+        try
+        {
             if (ActivationType == ToolActivationType.Protocol)
             {
-                result = await Launcher.LaunchUriAsync(new Uri($"{parsedArguments}"));
+                // Docs say this returns true if the default app for the URI scheme was launched;
+                // false otherwise. However, if there's no registered app for the protocol, it shows
+                // the "get an app from the store" dialog, and returns true. So we can't rely on the
+                // return value to know if the tool was actually launched.
+                var result = await Launcher.LaunchUriAsync(new Uri($"{parsedArguments}"));
+                if (result != true)
+                {
+                    // We get here if the user supplied a valid registered protocol, but the app failed to launch.
+                    var errorMessage = string.Format(
+                        CultureInfo.InvariantCulture, _errorMessageText, parsedArguments);
+                    throw new InvalidOperationException(errorMessage);
+                }
             }
             else
             {
                 if (ActivationType == ToolActivationType.Msix)
                 {
-                    var process = Process.Start("explorer.exe", $"shell:AppsFolder\\{AppUserModelId}");
-                    result = process is not null;
+                    process = LaunchPackagedTool(AppUserModelId);
                 }
                 else
                 {
-                    var startInfo = new ProcessStartInfo(Executable)
+                    string finalExecutable;
+                    string finalArguments;
+
+                    if (Executable.EndsWith(".msc", StringComparison.OrdinalIgnoreCase))
+                    {
+                        // Note: running most msc files requires elevation.
+                        finalExecutable = "mmc.exe";
+                        finalArguments = $"{Executable} {parsedArguments}";
+                    }
+                    else if (Executable.EndsWith(".ps1", StringComparison.OrdinalIgnoreCase))
+                    {
+                        // Note: running powershell scripts might need setting the execution policy.
+                        finalExecutable = "powershell.exe";
+                        finalArguments = $"{Executable} {parsedArguments}";

```

Comment:
nit: this will execute in the inbox version of powershell most likely. We may want to use Powershell (core) pwsh if its present (and give the user an option to select a path to powershell.

Reply:
Yes that's a good point. I'll leave more sophisticated functionality for a further round - there's probably other things we might want to consider. For example, I'm asserting that we intentionally don't help them set execution policy - but we might want to add that as a feature enhancement (or not, needs thinking about).